File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.java
class Solution {
    
    public int shortestPathBinaryMatrix(int[][] grid) {
        int r = grid.length, c = grid[0].length;
        int[][] directions = {{0,1},{1,0},{-1,0},{0,-1},
                             {1,1},{1,-1},{-1,1},{-1,-1}};
        LinkedList<int[]> q = new LinkedList<>();
        int[] start = new int[]{0, 0};
        if (grid[0][0] == 1) return -1;
        boolean[][] visited = new boolean[r][c];
        visited[0][0] = true;
        q.add(start);
        
        int steps = 1;
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int[] curr = q.poll();
                if (curr[0] == r - 1 && curr[1] == c - 1) return steps;
                for (int[] dir: directions) {
                    int nextI = curr[0] + dir[0];
                    int nextJ = curr[1] + dir[1];
                    if (nextI >= 0 && nextI < r && nextJ >= 0 && nextJ < c) {
                        if (!visited[nextI][nextJ] && grid[nextI][nextJ] == 0) {
                            visited[nextI][nextJ] = true;
                            int[] next =  new int[]{nextI, nextJ};
                            q.add(next);
                        } 
                    } 
                }
            }
            steps++;
        }
        
        return -1;
        
        
    }
}

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.cpp
class Solution {
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        int n = grid.size();
        if (grid[0][0] == 1 || grid[n - 1][n - 1] == 1) return -1;
        
        vector<vector<int>> directions = {{-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1}};
        
        queue<pair<int, int>> q;
        q.push({0, 0});
        grid[0][0] = 2; //mark as visited
        int steps = 1;
        
        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                pair<int, int> curr = q.front();
                q.pop();
                int r = curr.first, c = curr.second;
                if (r == n - 1 && c == n - 1) return steps;

                for (vector<int>& dir: directions) {
                    int next_r = r + dir[0], next_c = c + dir[1];

                    if (next_r >= 0 && next_c >= 0 && next_r < n && next_c < n && grid[next_r][next_c] == 0) {
                        
                        q.push({next_r, next_c});
                        grid[next_r][next_c] = 2;
                    }
                }
            }
            
            steps++;
        }
        
        return -1;
        
    }
};